testnet/src/burnchains/bitcoin_regtest_controller.rs

```diff
@@ -0,0 +1,699 @@
+                    break burnchain_tip;
+                }
+                sleep_ms(500);
+            }
+        }
+    }
+
+    fn submit_operation(&mut self, operation: BurnchainOperationType, op_signer: &mut BurnchainOpSigner) {
+
+        let transaction = match operation {
+            BurnchainOperationType::LeaderBlockCommit(payload) 
+                => self.build_leader_block_commit_tx(payload, op_signer),
+            BurnchainOperationType::LeaderKeyRegister(payload) 
+                => self.build_leader_key_register_tx(payload, op_signer),
+            BurnchainOperationType::UserBurnSupport(payload) 
+                => self.build_user_burn_support_tx(payload, op_signer)
+        };
+
+        let transaction = match transaction {
+            Some(tx) => SerializedTx::new(tx),
+            _ => return
+        };
+
+        self.send_transaction(transaction);
+    }
+    
+    #[cfg(test)]
+    fn bootstrap_chain(&mut self) {
+
+        if let Some(local_mining_pubkey) = &self.config.burnchain.local_mining_public_key {
+
+            let pk = hex_bytes(&local_mining_pubkey).expect("Invalid byte sequence");
+            let pkh = Hash160::from_data(&pk).to_bytes().to_vec();
+            let address = BitcoinAddress::from_bytes(
+                BitcoinNetworkType::Regtest,
+                BitcoinAddressType::PublicKeyHash,
+                &pkh)
+                .expect("Public key incorrect");
+
+            let request_builder = self.get_rpc_request_builder();
+
+            let _result = BitcoinRPCRequest::import_public_key(
+                request_builder, 
+                local_mining_pubkey);
+    
+            let request_builder = self.get_rpc_request_builder();
+            let result = BitcoinRPCRequest::generate_to_address(
+                request_builder, 
+                201,
+                address.to_b58());
+
+            match result {
+                Ok(_) => {},
+                Err(e) => {
+                    error!("Bitcoin RPC failure: error generating block {:?}", e);
+                    panic!();
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct SerializedTx {
+    bytes: Vec<u8>
+}
+
+impl SerializedTx {
+    pub fn new(tx: Transaction) -> SerializedTx {
+        let mut encoder = RawEncoder::new(Cursor::new(vec![]));
+        tx.consensus_encode(&mut encoder).expect("BUG: failed to serialize to a vec");
+        let bytes: Vec<u8> = encoder.into_inner().into_inner(); 
+        SerializedTx {
+            bytes
+        }
+    }
+
+    fn to_hex(self) -> String {
+        let formatted_bytes: Vec<String> = self.bytes.iter().map(|b| format!("{:02x}", b)).collect();
+        format!("{}", formatted_bytes.join(""))
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct UTXO {
+    pub txid: String,
+    pub vout: u32,
+    pub script_pub_key: String,
+    pub amount: f64,
```

Comment:
I'd stay away from using `f64` for encoding quantities of Bitcoin, if possible.  This can cause weird things to happen when dealing with small amounts of it.  Internally, Bitcoin represents all BTC as satoshis, for example.

Reply:
I'm using an `f64` because the `listunspent` rpc call is returning a quantity of expressed in BTC (even if the min sum amount is expressed in satoshi).
This `f64` is just used for un-marshalling the json, I can make this field private, it's wrapped by the method `get_sat_amount`.

Reply:
Can you please fix this before merging?

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1381/commits/5240beed0159868b1daf0e1b20bb5d631811c152